state-chain/chains/src/sol/api.rs

```diff
@@ -176,32 +201,94 @@ impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
 		// Lookup the current Aggkey
 		let agg_key = Environment::lookup_account(SolanaEnvAccountLookupKey::AggKey)?;
 		let (nonce_account, durable_nonce) = Environment::nonce_account()?;
 		let compute_price = Environment::compute_price()?;
 
-		// Build the instruction_set
-		let instruction_set = SolanaInstructionBuilder::transfer(
-			transfer_param,
-			agg_key,
-			nonce_account,
-			compute_price,
-		);
-
-		let transaction = SolTransaction::new_unsigned(SolMessage::new_with_blockhash(
-			&instruction_set,
-			Some(&agg_key.into()),
-			&durable_nonce.into(),
-		));
-
-		Ok(Self {
+		// split the transfer params into Native and Token assets
+		let (native, token) = transfer_param
+			.into_iter()
+			.partition::<Vec<TransferAssetParams<Solana>>, _>(|param| param.asset == SolAsset::Sol);
+
+		Ok(vec![
+			// Create native transfer instruction sets
+			native
+				.into_iter()
+				.map(|transfer_param| {
+					SolanaInstructionBuilder::transfer_native(
+						transfer_param.amount,
+						transfer_param.to,
+						agg_key,
+						nonce_account,
+						compute_price,
+					)
+				})
+				.collect::<Vec<_>>(),
+			if token.is_empty() {
+				vec![]
+			} else {
+				// Only do environment lookup if needed
+				let vault_program =
+					Environment::lookup_account(SolanaEnvAccountLookupKey::VaultProgram)?;
+				let vault_program_data_account = Environment::lookup_account(
+					SolanaEnvAccountLookupKey::VaultProgramDataAccount,
+				)?;
+				let token_vault_pda_account =
+					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenVaultPdaAccount)?;
+				let token_mint_pubkey =
+					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenMintPubkey)?;
+				let token_program_id = SolAddress::from_str(TOKEN_PROGRAM_ID)
+					.expect("Preset Solana Token program ID must be valid.");
+				let token_vault_ata = Environment::lookup_account(
+					SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount,
+				)?;
+				let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
+					.expect("Preset Solana System Program ID account must be valid.");
+
+				// Build the Transfer instruction set for Token transfers.
+				token
+					.into_iter()
+					.map(|transfer_param| {
+						let ata =
+						crate::sol::sol_tx_core::address_derivation::derive_associated_token_account(
+							transfer_param.to,
+							token_mint_pubkey,
+						)
+						.map_err(SolanaTransactionBuildingError::FailedToDeriveAddress)?;
+						Ok(SolanaInstructionBuilder::transfer_usdc_token(
+							ata.0,
+							transfer_param.amount,
+							transfer_param.to,
+							vault_program,
+							vault_program_data_account,
+							token_vault_pda_account,
+							token_vault_ata,
+							token_mint_pubkey,
+							token_program_id,
+							system_program_id,
+							agg_key,
+							nonce_account,
+							compute_price,
+						))
+					})
+					.collect::<Result<Vec<_>, _>>()?
+			},
+		]
+		.into_iter()
+		.flatten()
+		.map(|instruction_set| Self {
```

Comment:
This assumes that any asset in the FetchParams that is not Sol, is Usdc. I know we might not add any future tokens other than Usdc but would be nice for it to be extendible to other assets. 

In general, the way code is written assumes (also see fetch_from() function ) that when we are talking about token, we mean Usdc so it seems it wont be extendible to others potential tokens we might wanna support in the future

Reply:
I agree with that. It should be generalized so we can support any token by just passing a different `mint_pub_key` for the token.